src/skalo/read_graph.rs

```diff
@@ -0,0 +1,304 @@
+                for (exit_kmer, vec_variants) in tmp_container.iter() {
+                    // collect second to last kmer of each variant in a hashset -> test if at least 2 (ie, the variants end on a difference)
+                    let second_set: HashSet<u128> = vec_variants.iter().map(|v| v[1]).collect();
+                    let second_to_last_set: HashSet<u128> =
+                        vec_variants.iter().map(|v| v[v.len() - 2]).collect();
+
+                    if second_set.len() > 1 && second_to_last_set.len() > 1 {
+                        if let Some(most_common_length) = most_abundant_length(vec_variants) {
+                            let filtered_variants: Vec<_> = if vec_variants.len() == 2 {
+                                vec_variants.clone()
+                            } else {
+                                vec_variants
+                                    .iter()
+                                    .filter(|v| v.len() == most_common_length)
+                                    .cloned()
+                                    .collect()
+                            };
+
+                            let combined_ends: (u128, u128) = (*kmer, *exit_kmer);
+
+                            // build variants 1 by 1
+                            for vec_visited in filtered_variants {
+                                // build sequence
+                                let mut sequence = String::with_capacity(
+                                    vec_visited.len() + data_info.k_graph - 1,
+                                );
+                                sequence.push_str(&decode_kmer(*kmer, data_info.k_graph));
+                                let mut vec_snps: Vec<usize> = Vec::new();
+                                for (i, next) in vec_visited.iter().enumerate() {
+                                    if i != 0 {
+                                        // 1st corresponds to entry k-mer
+                                        sequence.push(get_last_nucl(*next));
+                                    }
+                                    if start_kmers.contains(next)
+                                        && i <= (vec_visited.len() - data_info.k_graph)
+                                    {
+                                        vec_snps.push(i + data_info.k_graph);
+                                    } else if end_kmers.contains(next) {
+                                        vec_snps.push(i - 1);
+                                    }
+                                }
+
+                                // save variant to container
+                                let variant =
+                                    VariantInfo::new(DnaSequence::encode(&sequence), vec_snps);
+
+                                tmp_container_2
+                                    .entry(combined_ends)
+                                    .or_default()
+                                    .push(variant);
```

Comment:
Might need to change this type depending on other changes, I've lost track of what `VariantInfo` and `DnaSequence` are